File: app/src/main/java/info/nightscout/androidaps/plugins/general/autotune/oref0/lib/percentile.js
// From https://gist.github.com/IceCreamYou/6ffa1b18c4c8f6aeaad2
// Returns the value at a given percentile in a sorted numeric array.
// "Linear interpolation between closest ranks" method
module.exports = function percentile(arr, p) {
    if (arr.length === 0) return 0;
    if (typeof p !== 'number') throw new TypeError('p must be a number');
    if (p <= 0) return arr[0];
    if (p >= 1) return arr[arr.length - 1];

    var index = arr.length * p,
        lower = Math.floor(index),
        upper = lower + 1,
        weight = index % 1;

    if (upper >= arr.length) return arr[lower];
    return arr[lower] * (1 - weight) + arr[upper] * weight;
}

File: core/utils/src/commonMain/kotlin/app/aaps/core/utils/Percentile.kt
package app.aaps.core.utils

import kotlin.math.floor

object Percentile {

    // Returns the value at a given percentile in a sorted numeric array, using the
    // "linear interpolation between closest ranks" method.
    // The rank index is arr.size * p, matching oref0 lib/percentile.js so that AndroidAPS autosens and
    // autotune reproduce OpenAPS dosing. This differs on purpose from the (arr.size - 1) * p variant in
    // the original https://gist.github.com/IceCreamYou/6ffa1b18c4c8f6aeaad2 - do NOT change it to n - 1,
    // it would shift every median and change autosens/autotune output.
    fun percentile(arr: Array<Double>, p: Double): Double = percentile(arr.toDoubleArray(), p)

    fun percentile(arr: DoubleArray, p: Double): Double {
        if (arr.isEmpty()) return 0.0
        if (p <= 0) return arr[0]
        if (p >= 1) return arr[arr.size - 1]
        val index = arr.size * p
        val lower = floor(index)
        val upper = lower + 1
        val weight = index % 1
        return if (upper >= arr.size) arr[lower.toInt()] else arr[lower.toInt()] * (1 - weight) + arr[upper.toInt()] * weight
    }
}